src/Compilers/CSharp/Test/Semantic/Utilities/ValueSetTests.cs

```diff
@@ -0,0 +1,663 @@
+            Assert.True(ForFloat.Related(Equal, float.NaN).Any(Equal, float.NaN));
+            Assert.Equal("Inf", ForDouble.Related(Equal, double.PositiveInfinity).ToString());
+            Assert.Equal("Inf", ForFloat.Related(Equal, float.PositiveInfinity).ToString());
+            Assert.Equal("-Inf", ForDouble.Related(Equal, double.NegativeInfinity).ToString());
+            Assert.Equal("-Inf", ForFloat.Related(Equal, float.NegativeInfinity).ToString());
+        }
+
+        [Fact]
+        public void TestDouble_04()
+        {
+            var neg = ForDouble.Related(LessThan, 0.0);
+            Assert.True(neg.Any(LessThan, double.MinValue));
+            Assert.False(neg.Any(GreaterThan, double.MaxValue));
+
+            var mi = ForDouble.Related(Equal, double.NegativeInfinity);
+            Assert.True(mi.All(LessThan, 0.0));
+        }
+
+        [Fact]
+        public void TestString_01()
+        {
+            var notaset = ForString.Related(Equal, "a").Complement();
+            var bset = ForString.Related(Equal, "b");
+            var intersect = bset.Intersect(notaset);
+            Assert.False(intersect.Any(Equal, "c"));
+        }
+
+        [Fact]
+        public void TestBool_Cov_01()
+        {
+            var t = ForBool.Related(Equal, true);
+            var f = ForBool.Related(Equal, false);
+            var em = t.Intersect(f);
+            Assert.True(em.IsEmpty);
+            var q = t.Intersect(t);
+            Assert.Same(t, q);
+            IValueSet b = t;
+            Assert.Same(b.Intersect(b), b);
+            Assert.Same(b.Union(b), b);
+            IValueSetFactory bf = ForBool;
+        }
+
+        [Fact]
+        public void TestByte_Cov_01()
+        {
+            var s = ForByte.Related(GreaterThan, 10).Intersect(ForByte.Related(LessThan, 100));
+            Assert.Equal("[11..99]", s.ToString());
+        }
+
+        [Fact]
+        public void TestString_Cov_01()
+        {
+            var s1 = ForString.Related(Equal, "a");
+            var s2 = ForString.Related(Equal, "b");
+            Assert.True(s1.Intersect(s2).IsEmpty);
+            Assert.True(s1.Complement().Union(s2.Complement()).Complement().IsEmpty);
+            Assert.Equal(s1.Union(s2).Complement(), s1.Complement().Intersect(s2.Complement()));
+            IValueSet b = s1;
+            Assert.Same(b.Intersect(b), b);
+            Assert.Same(b.Union(b), b);
+            IValueSetFactory bf = ForString;
+            Assert.False(s1.Union(s2).All(Equal, "a"));
+        }
+
+        [Fact]
+        public void TestDouble_Cov_01()
+        {
+            var s1 = ForDouble.Related(LessThan, 3.14d);
+            IValueSet b = s1;
+            Assert.Same(s1, s1.Intersect(s1));
+            Assert.Same(s1, s1.Union(s1));
+            var s2 = ForDouble.Related(GreaterThan, 31.4d);
+            var s3 = b.Complement().Intersect(s2.Complement());
+            Assert.Equal("NaN,[3.1400000000000001..31.399999999999999]", s3.ToString());
```

Comment:
>NaN [](start = 26, length = 3)

Does this mean `s3.Any(Equal, double.NaN) == true`? Shouldn't `s3.Any(Equal, double.NaN) == false` always, regardless of `s3`?

Reply:
No.  The pattern-matching of a constant pattern on a NaN input matches a NaN.  Everywhere else Equals is used in this implementation, it is intended that NaN should be treated as equal to itself.

---
In reply to: [383443458](https://github.com/dotnet/roslyn/pull/41503#discussion_r383443458) [](ancestors = 383443458)